src/editable_tree/mod.rs

```diff
@@ -584,45 +585,1164 @@ impl<'arena, Node: Ast<'arena>> DAG<'arena, Node> {
 }
 
 #[cfg(test)]
-mod integration {
-    use super::{cursor_path::CursorPath, EditSuccess, DAG};
+mod tests {
+    use super::{cursor_path::CursorPath, Direction, EditErr, EditSuccess, Side, DAG};
     use crate::arena::Arena;
-    use crate::ast::test_json::TestJSON as J;
+    use crate::ast::json::JSON;
+    use crate::ast::test_json::TestJSON;
     use crate::editor::normal_mode::Action;
 
+    fn run_test(
+        start_tree: TestJSON,
+        start_cursor_location: CursorPath,
+        action: Action,
+        expected_edit_result: (bool, Result<EditSuccess, EditErr>),
+        expected_tree: TestJSON,
+        expected_cursor_location: CursorPath,
+    ) {
+        let arena: Arena<JSON> = Arena::new();
+        let root = start_tree.add_to_arena(&arena);
+        let mut editable_tree = DAG::new(&arena, root, start_cursor_location);
+
+        assert_eq!(
+            expected_edit_result,
+            editable_tree.execute_action(action),
+            "Not equal in action result"
+        );
+        assert_eq!(expected_tree, editable_tree.root(), "Not equal in tree.");
+        assert_eq!(
+            expected_cursor_location, editable_tree.current_cursor_path,
+            "Not equal in cursor location."
+        );
+    }
+
+    #[test]
+    fn dag_root_insertchild() {
+        let start_tree = TestJSON::Array(vec![]);
+        let start_cursor_location = CursorPath::from_vec(vec![]);
+        let action = Action::InsertChild('f');
+
+        let expected_edit_result: (bool, Result<EditSuccess, EditErr>) = (
+            false,
+            Ok(EditSuccess::InsertChild {
+                c: 'f',
+                name: "false".to_string(),
+            }),
+        );
+        let expected_tree = TestJSON::Array(vec![TestJSON::False]);
+        let expected_cursor_location = CursorPath::from_vec(vec![]);
+
+        run_test(
+            start_tree,
+            start_cursor_location,
+            action,
+            expected_edit_result,
+            expected_tree,
+            expected_cursor_location,
+        );
```

Comment:
I think we should probably inline the unit tests - so pass the parameters directly to `run_test`.  So this test would end up as
```rust
run_test(
    TestJSON::Array(vec![]),
    CursorPath::from_vec(vec![]),
    Action::InsertChild('f'),
    (
        false,
        Ok(EditSuccess::InsertChild {
            c: 'f',
            name: "false".to_string(),
        }),
    ),
    TestJSON::Array(vec![TestJSON::False]),
    CursorPath::from_vec(vec![]),
);
```

which I think is much more compact and readable.